components/places/src/storage.rs

```diff
@@ -248,6 +251,104 @@ fn add_visit(db: &impl ConnExt,
     Ok(RowId(rid))
 }
 
+// Support for Sync - in its own module to try and keep a delineation
+pub mod history_sync {
+    use super::*;
+    use sync::history::record::{HistoryRecordVisit};
+
+    #[derive(Debug)]
+    pub struct FetchedVisit {
+        pub is_local: bool,
+        pub visit_date: Timestamp,
+        pub visit_type: Option<VisitTransition>,
+    }
+
+    impl FetchedVisit {
+        pub fn from_row(row: &Row) -> Result<Self> {
+            Ok(Self {
+                is_local: row.get_checked("is_local")?,
+                visit_date: row.get_checked::<_, Option<Timestamp>>("visit_date")?
+                     .unwrap_or_default(),
+                visit_type: VisitTransition::from_primitive(
+                                row.get_checked::<_, Option<u8>>("visit_type")?.unwrap_or(0)
+                            ),
+            })
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct FetchedVisitPage {
+        pub url: Url,
+        pub guid: SyncGuid,
+        pub row_id: RowId,
+        pub title: String,
+    }
+
+    impl FetchedVisitPage {
+        pub fn from_row(row: &Row) -> Result<Self> {
+            Ok(Self {
+                url: Url::parse(&row.get_checked::<_, Option<String>>("url")?.expect("non null column"))?,
+                guid: SyncGuid(row.get_checked::<_, Option<String>>("guid")?.expect("non null column")),
+                row_id: row.get_checked("id")?,
+                title: row.get_checked::<_, Option<String>>("title")?.unwrap_or_default(),
+            })
+        }
+    }
+
+    pub fn fetch_visits(db: &Connection, url: &Url, limit: u32) -> Result<Option<(FetchedVisitPage, Vec<FetchedVisit>)>> {
+        // We do this in 2 steps - "do we have a page" then "get visits"
+        let page_sql = "
+          SELECT guid, url, id, title
+          FROM moz_places h
+          WHERE url_hash = hash(:url) AND url = :url";
+
+        let page_info = match db.try_query_row(page_sql, &[(":url", &url.clone().into_string())], FetchedVisitPage::from_row, true)? {
+            None => return Ok(None),
+            Some(pi) => pi,
+        };
+
+        let mut stmt = db.prepare("
+          SELECT is_local, visit_type, visit_date
+          FROM moz_historyvisits
+          WHERE place_id = :place_id
+          LIMIT :limit")?;
+        let iter = stmt.query_and_then_named(&[(":place_id", &page_info.row_id),
+                                               (":limit", &limit)],
+                                             FetchedVisit::from_row)?;
+        // markh should work out how to iter.collect this ;)
```

Comment:
You can collect from a `Iterator<Item = Result<T>>` into a `Result<Vec<T>>`. e.g. `let mut visits = iter.collect::<Result<Vec<_>>>()?` might do it (you might need to use `Result<Vec<_>, _>` instead, it depends on whats in scope and etc).